hbase-server/src/main/java/org/apache/hadoop/hbase/master/procedure/TableProcedureWaitingQueue.java

```diff
@@ -0,0 +1,113 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.hadoop.hbase.master.procedure;
+
+import java.util.ArrayDeque;
+import java.util.Optional;
+import java.util.Queue;
+import java.util.function.Function;
+import org.apache.hadoop.hbase.procedure2.Procedure;
+import org.apache.yetus.audience.InterfaceAudience;
+
+import org.apache.hbase.thirdparty.com.google.common.base.Preconditions;
+
+/**
+ * To prevent multiple Create/Modify/Disable/Enable table procedures run at the same time, we will
+ * keep table procedure in this queue first before actually enqueuing it to
+ * MasterProcedureScheduler's tableQueue. See HBASE-28683 for more details
+ */
+@InterfaceAudience.Private
+class TableProcedureWaitingQueue {
+
+  private final Function<Long, Procedure<?>> procedureRetriever;
+
+  // whether there is already a table procedure enqueued in ProcedureScheduler.
+  private Procedure<?> enqueuedProc;
+
+  private final Queue<Procedure<?>> queue = new ArrayDeque<>();
+
+  TableProcedureWaitingQueue(Function<Long, Procedure<?>> procedureRetriever) {
+    this.procedureRetriever = procedureRetriever;
+  }
+
+  private boolean isSubProcedure(Procedure<?> proc) {
+    while (proc.hasParent()) {
+      if (proc.getParentProcId() == enqueuedProc.getProcId()) {
+        return true;
+      }
+      proc = Preconditions.checkNotNull(procedureRetriever.apply(proc.getParentProcId()),
+        "can not find parent procedure pid=%s", proc.getParentProcId());
+    }
+    return false;
+  }
+
+  /**
+   * Return whether we can enqueue this procedure to ProcedureScheduler.
+   */
```

Comment:
Here we can also mention that if it returns false, it also means that the proc is queued up for waiting purpose.